### ISSUE TITLE
[no-invalid-void-type] allowInGenericTypeArguments option not working as expected

### ISSUE THREAD

**Issue by mure** [typescript-eslint/typescript-eslint#3143]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-invalid-void-type": "error"
  }
}
```
or
```JSON
{
  "rules": {
    "@typescript-eslint/no-invalid-void-type": ["error", { "allowInGenericTypeArguments": true }]
  }
}
```
```TS
switchMap((webapp: WebApp): Observable<[Document, void]> => combineLatest([this.getDocumentFromWebApp(webapp), this.loadDocuments(this.activePlugin)])),
```

**Expected Result**

I would expect this not to error because 'void' is only used as a generic type argument for the Observable.

**Actual Result**

`error    void is only valid as a return type or generic type variable                     @typescript-eslint/no-invalid-void-type`

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.16.1` |
| `@typescript-eslint/parser`        | `3.9.0` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.6.0` |
| `node`                             | `X.Y.Z` |

**Comment by bradzacher** [comment]
> I would expect this not to error because 'void' is only used as a generic type argument for the Observable

It's not actually in a generic type argument.
It's used in a tuple type, and that tuple type is used in a generic type argument.

To show what I mean - code can be written like so without changing its meaning:
```ts
type Foo = [Document, void];
type T = Observable<Foo>;
```

This code is obviously a violation of the rule.

I don't know how your code works, but based on the type, I'm assuming:
```ts
switchMap(
  (webapp: WebApp): Observable<[Document, void]> =>
    combineLatest([
      this.getDocumentFromWebApp(webapp), // return type === Document
      this.loadDocuments(this.activePlugin), // return type === void
    ]),
)
```

Which means that it's actually returning an `Observable` which returns an array (tuple) whose second element will always be the value `undefined`.
So in this instance you should use the `undefined` type.
